### ISSUE TITLE
Derivations built with writableStoreOverlay and reboots

### ISSUE THREAD

**Issue by Eisfunke** [astro/microvm.nix#134]
I've added a `writableStoreOverlay`to a microvm according to the instructions here: https://astro.github.io/microvm.nix/shares.html

This does work, I can now build/download derivations inside the vm, e.g. with `nix run nixpkgs#hello`. However, when the vm is rebooted, and try the same command again, the derivations are downloaded again.

As far as I understand, that happens because the Nix daemon doesn't know about those derivations anymore, because its database isn't persisted acroos the reboot. So it won't know about previously added derivations, even though the paths still exist in the filesystem.

As I understand the code, this problem is tackled for the paths needed for the vm's system config via `nix-store --load-db` during boot, see [here](https://github.com/astro/microvm.nix/blob/main/nixos-modules/microvm/store-disk.nix).

Is my understanding correct and this is expected? If so, I think this should be explained in the docs (I can prepare a PR for that), because I intuitively expected writableStoreOverlay to be usable across reboots.

Also, is there a proper way to get a writable store in a microvm that survives reboots and preferably enables garbage collection? It seems microvms always either have a read-only squashfs store with optionally the writable overlay, or if you turn `storeOnDisk` off wants to use the host store directly, and it seems like a very bad idea to try and write stuff there from inside the vm. I also tried mounting `/nix` or `/nix/var` as a share additionally to using the writable overlay, but that just seems to corrupt after rebooting.

(My, admittedly convoluted, usecase: I'm trying to run a GitLab Runner that can build Nix derivations inside of a microvm for additional isolation and so that I don't have to have a Docker daemon running on my host system. I'd prefer to be able to keep already built derivations cached across reboots. My config for reference: https://git.eisfunke.com/config/nixos/-/blob/main/nixos/server/git.nix)

Thank you for this project! Sorry for the long issue, I'd be thankful for any pointers in the right direction.

**Comment by astro** [comment]
Your reckoning is 100% right. I agree, I should add that behaviour to the docs until the problem is solved. It would be nice to keep the derivations across reboots, so feel encouraged to research further!

**Comment by vector1dev** [comment]
Is any workaround available?

**Comment by Eisfunke** [comment]
What I'm currently trying is using a self-hosted binary cache: https://github.com/zhaofengli/attic

You can push any derivations built in the VM to the store and won't have to re-build them. As you can host the cache on the same machine or network it should be pretty quick to download from there. Could be an alternative if you don't strictly need the derivations to stay in the VM's store, I'd even argue it's cleaner because the VM needs less state that way.

**Comment by mksafavi** [comment]
Hi @astro 
Is this issue still open? I have the same problem.
Also I think this is not mentioned in the docs.

```nix
  config.microvm.writableStoreOverlay = "/nix/.rw-store";
  config.microvm.shares = [
    {
      tag = "ro-store";
      source = "/nix/store";
      mountPoint = "/nix/.ro-store";
    }
  ];
  config.microvm.volumes = [
    {
      image = "nix-store.img";
      mountPoint = "/nix/.rw-store";
      size = config.mariner.storage.nixStoreSizeMiB;
    }
  ];
```

**Comment by SuperSandro2000** [comment]
The caveat in https://github.com/microvm-nix/microvm.nix/commit/25460d97cffffe24f565a9d654c4aab1848598b4 is still true to my knowledge.

**Comment by mksafavi** [comment]
> The caveat in [25460d9](https://github.com/microvm-nix/microvm.nix/commit/25460d97cffffe24f565a9d654c4aab1848598b4) is still true to my knowledge.

Right. I didn't see that note. Just to check that I understand it correctly, by "delete and recreate" it means that you will lose the new derivations built inside the VM. right? 


I noticed that the issue is due to missing the db, So I added this to my config. I haven't tested it that much but it seems to work. At least when I `nix-shell -p hello` it survives the reboot.

```
  config.fileSystems = {
    "/nix/var" = {
      device = "/nix/.rw-store/var";
      options = [ "bind" ];
      fsType = "none";
      depends = [ "/nix/.rw-store" ];
    };
  };
  config.systemd.tmpfiles.rules = [
    "d /nix-store/var 0755 root root -"
  ];
```

mounted paths:
```
[vm@nixos:~]$ mount | grep -E '(/nix|/nix/var)'
ro-store on /nix/.ro-store type 9p (rw,relatime,access=client,msize=65536,trans=virtio,x-initrd.mount,x-systemd.after=systemd-modules-load.service)
/dev/vdb on /nix/.rw-store type ext4 (rw,relatime,x-initrd.mount)
overlay on /nix/store type overlay (rw,relatime,lowerdir=/sysroot/nix/.ro-store,upperdir=/sysroot/nix/.rw-store/store,workdir=/sysroot/nix/.rw-store/work,x-initrd.mount,x-systemd.requires-mounts-for=/sysroot/nix/.ro-store,x-systemd.requires-mounts-for=/sysroot/nix/.rw-store/store,x-systemd.requires-mounts-for=/sysroot/nix/.rw-store/work)
overlay on /nix/store type overlay (ro,nosuid,nodev,relatime,lowerdir=/sysroot/nix/.ro-store,upperdir=/sysroot/nix/.rw-store/store,workdir=/sysroot/nix/.rw-store/work)
/dev/vdb on /nix/var type ext4 (rw,relatime,x-systemd.requires-mounts-for=/nix/.rw-store)
```

**Comment by SuperSandro2000** [comment]
The DB needs to be updated with new store contents on updates and it generally does not work when the store from the host is shared. This is a hard problem to solve.
